ci-operator/step-registry/ipi/conf/aws/custom-vpc/ipi-conf-aws-custom-vpc-commands.sh

```diff
@@ -1,49 +1,200 @@
+if [ ! -f "${pub_subnets_file}" ] || [ ! -f "${priv_subnets_file}" ] || [ ! -f "${az_file}" ]; then
+  echo "File ${pub_subnets_file} or ${priv_subnets_file} or ${az_file} does not exist."
   exit 1
 fi
 
-echo -e "subnets: $(cat ${subnet_ids_file})"
+ZONE_COUNT=$(cat "${az_file}" | jq '.|length')
+
+echo -e "public subnets: $(cat ${pub_subnets_file})"
+echo -e "private subnets: $(cat ${priv_subnets_file})"
 echo -e "AZs: $(cat ${az_file})"
 
-CONFIG_SUBNETS="${SHARED_DIR}/install-config-subnet.yaml.patch"
-cat > "${CONFIG_SUBNETS}" << EOF
-platform:
-  aws:
-    subnets: $(cat "${subnet_ids_file}")
-EOF
-yq-go m -x -i "${CONFIG}" "${CONFIG_SUBNETS}"
 
+function public_only()
+{  
+  if [[ "${OPENSHIFT_INSTALL_AWS_PUBLIC_ONLY:-}" != "" ]]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+function private_cluster()
+{  
+  if [ -n "${PUBLISH}" ] && [ X"${PUBLISH}" == X"Internal" ]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+function patch_az()
+{
+    local config=$1
+    shift
+    azs=$(echo "$@" | yq-v4 -o yaml 'split(" ") | map(.)')
+    export azs
+    yq-v4 eval -i '.compute[0].platform.aws.zones += env(azs)' ${config}
+    yq-v4 eval -i '.controlPlane.platform.aws.zones += env(azs)' ${config}
+    unset azs
+}
+
+function patch_legcy_subnets()
+{
+    local config=$1
+    shift
+    for subnet in "$@"; do
+        export subnet
+        yq-v4 eval -i '.platform.aws.subnets += [env(subnet)]' ${config}
+        unset subnet
+    done
+}
+
+function patch_new_subnets()
+{
+    local config=$1
+    shift
+    for subnet in "$@"; do
+        export subnet
+        yq-v4 eval -i '.platform.aws.vpc.subnets += [{"id": env(subnet)}]' ${config}
+        unset subnet
+    done
+}
+
+function patch_new_subnet_with_roles()
+{
+    local config=$1
+    local subnet=$2
+    shift 2
+    if [[ "$*" == "" ]]; then
+	    return 0
+    fi
+    roles=$(echo "$@" | yq-v4 -o yaml 'split(" ") | map({"type": .})')
+    export subnet roles
+    yq-v4 eval -i '.platform.aws.vpc.subnets += [{"id": env(subnet), "roles": env(roles)}]' ${config}
+    unset subnet roles
+}
 
 if [[ ${ADD_ZONES} == "yes" ]]; then
-  CONFIG_AZ="${SHARED_DIR}/install-config-azs.yaml.patch"
-  cat > "${CONFIG_AZ}" << EOF
-controlPlane:
-  platform:
-    aws:
-      zones: $(cat "${az_file}")
-compute:
-- platform:
-    aws:
-      zones: $(cat "${az_file}")
-EOF
-  yq-go m -x -i "${CONFIG}" "${CONFIG_AZ}"
+  patch_az $CONFIG $(jq -r '.|join(" ")' "${az_file}")
+fi
+
+if ((ocp_major_version == 4 && ocp_minor_version <= 18)); then
+  if private_cluster; then
+    echo "This is a private cluster so use only private subnets from the VPC"
+    patch_legcy_subnets "${CONFIG}" $(jq -r '.|join(" ")' "${priv_subnets_file}")
+  else
+    patch_legcy_subnets "${CONFIG}" $(jq -r '.|join(" ")' "${priv_subnets_file}")
+    patch_legcy_subnets "${CONFIG}" $(jq -r '.|join(" ")' "${pub_subnets_file}")
+  fi
+else
+  # for 4.19+
+  if [[ ${ASSIGN_ROLES_TO_SUBNETS} == "yes" ]]; then
```

Comment:
Can directly judge `SUBNET_ROLES` empty or not to save `ASSIGN_ROLES_TO_SUBNETS` var ?

Reply:
`ASSIGN_ROLES_TO_SUBNETS` is easier for _reading_ in specific steps, like `cucushift-installer-check-aws-custom`, and, for local zone subnets, we usually generate one subent, private or private, its a little hard to config `SUBNET_ROLES` for such test, so `ASSIGN_ROLES_TO_SUBNETS` is good choice for `ipi-conf-aws-edge-zone`